### ISSUE TITLE
Syntax is Style philosophy?

### ISSUE THREAD

**Issue by Qix-** [arua-lang/proposal#17]
With Python, we have [PEP8](https://www.python.org/dev/peps/pep-0008/) which is a 'governmental' style guide imposed (or at least heavily pushed) by the Python org itself. It seems to its job of making others look poorly upon you if you don't use it.

This brings to question, should a language impose a style on its users? My compulsory response says "no" but thinking about it further I have concluded it's really up to the language.

For example, there are a lot of subjective things about braced languages: do braces go on the same line, or the next line? Do closing braces go on the last line or on their own line? These are things often enforced by a style guide or a linter rather than the compiler itself.

What if that wasn't the case?

I feel like a lot of decisions Arua is making about the language itself are resulting in fairly clean syntax. A lot of symbols not essential to the language have been taken out. An example are function definitions:

``` arua
fn someFunction(arg1 i32, arg2 str) f32
    # ...
```

In the above example, _there isn't much to really argue about_ - it boils down to just a few things:
- space before the `(` after `someFunction`? 
- space after `#` in the comment?
- tabs, or spaces?
- how many tabs or spaces?

While these four things are important, quantitatively these are much less than other languages.

With the option disabling it, should Arua enforce a style by default?

**Comment by Qix-** [comment]
// @sindresorhus @iankronquist @Polyg0n

**Comment by chrmatique** [comment]
I feel like there should be at least a minimal style guide for anybody that writes code for the main arua packages/libraries/etc. What they want to do with arua-lang code for their own packages? I could care less.

As far as having the style in your example? You'd want to make sure users write all params on a single line, unlike what some C++ programmers do:

``` c++
 bool getNode(
        bool hasPriorityLoad,
        long nodeUID,
        char nodeName,
        unsigned int loadOrder,
        node_loc_t nodeLocation
     );
```

Due to your language being a whitespace/carriage return based language (like python) you wouldn't want to do this for obvious reasons.

As far as comments.. Putting a space after the comment token helps with readability.
Tabbing and spacing? I would impose a 4-space ruling in the language, it's a pretty well used standard in programming, and really makes the language more readable. And spacing, no tabs... Because hard tabs break the spacing when you push it to github (Believe me I've tried it).

And if you have an option to disable it? I'd just impose the style standard by default. Especially if someone is uploading their code to a main arua-lang repo, just so everyone is using the same styling.

**Comment by Qix-** [comment]
I'm about to add a placeholder for a scoping RFC, but ultimately indentation is tabs first, then spaces (no interleaving) - and the tuple of those counts becomes the indent. Tabs take precedence.

| row <=> col | 1 space | 1 tab | 1 tab, 1 space | 1 tab, 2 spaces | 2 tabs, 1 space | 2 tabs, 2 spaces |
| --- | --- | --- | --- | --- | --- | --- |
| 1 space | = | < | < | < | < | < |
| 1 tab | > | = | < | < | < | < |
| 1 tab, 1 space | > | > | = | < | < | < |
| 1 tab, 2 spaces | > | > | > | = | < | < |
| 2 tabs, 1 space | > | > | > | > | = | < |
| 2 tabs, 2 spaces | > | > | > | > | > | = |

**Tabs are to be disallowed anywhere outside indentation**, following the dogma of _tabs for indentation, spaces for alignment_. As a side effect, this allows users that prefer space indentation and _absolutely no tabs at all_ to do just that, without conflict.

Everyone wins.

---

You're right about newlines in argument lists. They make no sense in Arua and frankly they're ugly in languages that allow them (purely opinion). @sindresorhus has pointed out a few times we're not in a time where IDE widths are a problem anymore, so it makes a bit of sense from a _functional_ standpoint, too.

---

The space after `#` in comments should be a stylistic error, which leads me to want to have a specific standards about what types of errors there are (errors, warnings, and style notices, etc.) and differentiate between them in the compiler output.

**Comment by sindresorhus** [comment]
A language should make you more productive. Bikeshedding about style inflicts productivity. I've personally experienced this. If I ever were to design a language, I would be very strict about style.

**Comment by chrmatique** [comment]
@Qix- 
I agree with @sindresorhus. I would really think about enforcing a strict style to maintain clean code.
